tests/MongoDB.Driver.Core.Tests/Core/Operations/AsyncCursorTests.cs

```diff
@@ -313,143 +311,151 @@ private ConnectionDescription CreateConnectionDescriptionSupportingSession()
                 new MessageEncoderSettings(),
                 maxTime.WithDefault(null));
         }
+    }
 
-        // nested types
-        private class Reflector
+    public class AsyncCursorIntegrationTests : OperationTestBase
+    {
+        [Theory]
+        [InlineData(0, 1000)]
+        [InlineData(2, 2)]
+        [InlineData(2, 1000)]
+        [InlineData(4, 2)]
+        [InlineData(4, 4)]
+        [InlineData(4, 1000)]
+        public void Session_reference_count_should_be_decremented_as_soon_as_possible(int collectionSize, int batchSize)
         {
-            // private fields
-            private AsyncCursor<BsonDocument> _instance;
+            RequireServer.Check();
+            DropCollection();
+            var documents = Enumerable.Range(1, collectionSize).Select(n => new BsonDocument("_id", n));
+            Insert(documents);
 
-            // constructors
-            public Reflector(AsyncCursor<BsonDocument> instance)
+            var findOperation = new FindOperation<BsonDocument>(_collectionNamespace, BsonDocumentSerializer.Instance, _messageEncoderSettings)
             {
-                _instance = instance;
-            }
+                BatchSize = batchSize
+            };
 
-            // public properties
-            public int? BatchSize
+            using (var binding = new WritableServerBinding(CoreTestConfiguration.Cluster, _session.Fork()))
+            using (var cursor = findOperation.Execute(binding, CancellationToken.None))
             {
-                get
+                AssertExpectedChannelSourceDisposedValue(cursor);
```

Comment:
Because of the weird way sessions are implemented (part in High, part in Core, and part in glue code that straddles the boundary between them) there is no single reference count we can assert against.

But what we can do is assert that the channel source is disposed as soon as possible, and assume that in the absence of bugs in the channel source (tested separately) that would result in the server session being returned to the pool as soon as possible.